Fix two coverity scan reported issues

tonycoz opened the pull request:
1) a loop with an off-by-one ending index

2) uninitialized structure members being passed along to standard library functions

tonycoz committed:
stop coverity complaining about copying uninitialised tm_gmtoff, tm_zone

This was reported in the two callers to S_ints_to_tm():

Using uninitialized value mytm.tm_gmtoff when calling S_ints_to_tm.
Using uninitialized value mytm.tm_zone when calling S_ints_to_tm.

cids 464354, 464357

tonycoz committed:
regdump_intflags: there is no  PL_reg_intflags_name[REG_INTFLAGS_NAME_SIZE]

Yes, this is an old report, from 2014.

cid29034

mauke commented on regcomp_debug.c:
```diff
@@ -320,7 +320,7 @@ S_regdump_intflags(pTHX_ const char *lead, const U32 flags)
 
     ASSUME(REG_INTFLAGS_NAME_SIZE <= sizeof(flags)*8);
```
This line doesn't make sense to me. Both `REG_INTFLAGS_NAME_SIZE` and `sizeof flags` are compile-time constants. There is no need for the compiler to "assume" anything to optimize the code; it already knows all the values involved. This is just an `assert`.

(And for maximum portability, this `8` should be `CHAR_BIT` instead.)

khwilliamson replied:
Off-topic, concerning `CHAR_BIT`.  Without looking, this appears to be defined starting in C99.  We have it defined ourselves in places like pp_pack.c, and there is `CHARBITS`  from Configure, which could be removed and replaced with `CHAR_BIT`

mauke replied:
`CHAR_BIT` is from C89, actually: https://en.cppreference.com/w/c/types/limits

tonycoz replied:
It's not directly related to either of the fixes here, but it is in the spirit of the little changes in this PR.

I've changed that `ASSUME()` and another in the same file to `STATIC_ASSERT_STMT()`.

Perl in general uses `CHARBITS` so I followed along with that rather than using `CHAR_BIT`.

Other uses of ASSUME seem reasonable, though I haven't looked at them in detail.

mauke replied:
I see your commit, but not in this pull request. Did you accidentally commit to the wrong branch?

It looks like there are 3 branches:

1. `Perl:blead`, the target of this PR
2. `tonycoz:cid464357-using-uninitialized`, the source of this PR, which contains the `struct tm` and off-by-one fixes
3. `Perl:cid464357-using-uninitialized`, which is like [#2](https://github.com/tonycoz/perl5/tree/cid464357-using-uninitialized) but also contains the `STATIC_ASSERT_STMT` changes

[Branch #3](https://github.com/Perl/perl5/tree/cid464357-using-uninitialized) is unrelated to this PR because it exists in the Perl repository, not your fork of it (like [branch #2](https://github.com/tonycoz/perl5/tree/cid464357-using-uninitialized)).

tonycoz replied:
Doh,  it's not the first time I've made that mistake when pushing from `git gui`.  Thanks and fixed.

tonycoz committed:
regcomp_debug.c: change a mis-use of ASSUME() to STATIC_ASSERT_STMT()

As mauke points out in #21490 the compiler "knows" all of the
constants here at compile-time, so the ASSUME()s here do nothing.

I suspect the intent here was to ensure that we never overflow the
capacity of the flags parameter, which we can check at compile-time
with STATIC_ASSERT_STMT().

mauke also suggested using CHAR_BIT instead of 8, but Configure
probes for CHARBITS and the perl core generally uses CHARBITS,
so use that here too.
